### PR TITLE
allow disabling default keymaps

### PR DESCRIPTION
**Pull request by osleg**
In response to https://github.com/ggandor/lightspeed.nvim/issues/11 
This PR allowing to *NOT MAP AT ALL* some of the mappings
It separates mapping to 3 categories and adding 3 options to the `setup`
1. `override-s` - default `true` overrides `s`
2. `override-x`- default `true` overrides `x`
3. `override-motion` - default `true` overrides `fFtT...`

### CODE REVIEW

**Comment by kutsan** [comment]
Friendly ping @ggandor.

**Comment by ggandor** [comment]
Hi! I'm a bit exhausted, and taking some time off for a couple of weeks, and I would especially not want to deal with API-related issues and breaking changes now. ;) Thanks for understanding.

That said, if such an option would ever be provided, it would be one flag, something like `set_default_keymaps = true`, this level of granularity is unnecessary. (On a side note, I don't understand the separate `override_x` option, AFAIK, `x` is not used in operator-pending mode by Vim. Is it?)

**Comment by osleg** [comment]
@ggandor oh I didn't expect your attention to the PR, I made it as response to someone above asking for it but saying they don't know fennel, so I figured since I need that change myself why wouldn't I make it public? :)

As per the separation of `override_<thing>` - it comes back again to this thread. Some people, like me, dont use the `x` binding but we do use the original `x` behaviour. Same goes for `s` and `fFtT...`, I personally want to keep motions in their original form, so the only thing I care about is the `s` binding. I also figured not binding `fFtT...` would be useful for those who heavily rely on macros... 
Bottom line - go figure what people want, the more flexible it is - the better (in specific case of vim plugins), since we have all flavours of users.

**Comment by ggandor** [comment]
> the original `x` behaviour

What is e.g. `yx...` supposed to do?

>  I also figured not binding fFtT... would be useful for those who heavily rely on macros...

That is a typical XY problem, I even labeled #11 as such, because of this. Wouldn't it be more logical for those people to contribute to the conversation in #14 instead? The very point of the plugin is to enhance the native `ft`, lightspeed motions (`f/t/s/x`) together form a suite.

> the more flexible it is - the better

It's not that simple. Flexibility is great while it comes for free, but that is rarely the case. Having one way of doing things, conventions people - including other plugin authors - can rely on, a small, non-cluttered API, etc. can be features too. Plugins like Lightspeed/Sneak/Surround/etc. are kinda special, as they provide new/enhanced _basic_ motions (/text objects), and not just a bunch of special commands for, say, git stuff, window handling, code formatting, or anything else. They aim to integrate into the very core of Vim as a text editor.

> I didn't expect your attention to the PR, I made it as response to someone in #11 asking for it but saying they don't know fennel, so I figured since I need that change myself why wouldn't I make it public? :)

Right, I suppose it's OK to close this then. (Sorry, but even if this is just a quick hack together, I cannot help mentioning the wtf factor of those 3 copy-pasted loops... DRY? :D)

Fennel note: `(:foo bar)` only works in Clojure. Semantically `:foo` is a normal Lua string in Fennel, mere syntax sugar for `"foo"`. The only correct way is thus `(. bar :foo)`.

**Comment by osleg** [comment]
> Fennel note: (:foo bar) only works in Clojure. Semantically :foo is a normal Lua string in Fennel, mere syntax sugar for "foo". The only correct way is thus (. bar :foo).

Thanks for the info, I felt there's something off

> What is e.g. `yx...` supposed to do?

Not even a slightest idea. `x` by itself removing character under the cursor. And my muscle memory of 22 years disagrees with such change

> That is a typical XY problem, I even labeled #11 as such, because of this. Wouldn't it be more logical for those people to contribute to the conversation in #14 instead? 

No, as I don't rely on macros much, because you took it a bit off the context. And I can't talk for 'those people'
 
The main context was that I personally don't like changed behaviour of `motions`. I don't like them highlighted, don't like them span on multiple lines and overall find them useless for myself. The `s` binding is answering all of my needs for fast navigation and it does it awesomely, everything else however (for me) is your subjective opinion.


> Plugins like Lightspeed/Sneak/Surround/etc. are kinda special, as they provide new/enhanced basic motions (/text objects)
or 

Here's another opinion for you: all this plugins are really doing (except being arguably useful) is not following the common path of giving the power to the end user and making opinionated decisions on user's behalf which, for some users at beginning and for most users as they advance those plugins become red flag. And in the end `surround` got removed from plugins list forever.

> Right, and thanks - I suppose it's OK to close this then(?)

That's up to you, be it my repo and would I not want to implement such a change I would keep it open for those who seek.

**Comment by ggandor** [comment]
> yeah i don't think i ever used x in OP

Okay, I'm lost here. Why make an API flag if you don't even _know_ whether it's used by Vim? (Why didn't you create an `override_z` flag then?)

> making opinionated decisions on user's behalf which, for some users at beginning and for most users as they advance those plugins become red flag

https://github.com/ggandor/lightspeed.nvim#guiding-principles A balance has to be found between providing flexibility and tools that "just work". If every plugin would follow the "start from scratch and do whatever you want, I'm not making any assumptions, I'm just a DIY toolkit" approach, them most of us would go insane. But there's a difference in explicitly disallowing something and not making it _easy_. Lightspeed is designed as a suite of motions, and people who do not want the whole pack are not handled as first-class citizens, that's all. It is a design decision.

**Comment by ggandor** [comment]
Closing in favor of 61b1afa4eb9105803b583e88cda53385f80ebca1.
